### PR TITLE
Move more ArrayBuffer logic into ArrayStorage.

### PR DESCRIPTION
**Pull request by dabrahams**
This is a step along the way to AnyArrayBuffer.  Also updated some docs.

### CODE REVIEW

File: Sources/PenguinStructures/ArrayStorage.swift [hunk 1]
```diff
@@ -62,13 +85,30 @@ public protocol AnyArrayStorageImplementation: AnyArrayStorage {
   /// Appends the instance of the concrete element type whose address is `p`,
   /// returning the index of the appended element, or `nil` if there was
   /// insufficient capacity remaining
+  ///
+  /// Complexity: O(1)
```
Comment (saeta) [inline Sources/PenguinStructures/ArrayStorage.swift:89]: Nit:

```suggestion
  /// - Complexity: O(1)
```

File: Sources/PenguinStructures/ArrayStorage.swift [hunk 2]
```diff
@@ -122,7 +162,58 @@ extension ArrayStorageImplementation {
...
+  /// Returns a copy of `self` after appending `x`, moving elements from the
+  /// existing storage iff `moveElements` is true.
+  ///
+  /// - Postcondition: if `count == capacity` on invocation, the result's
+  ///   `capacity` is `self.capacity` scaled up by a constant factor.
+  ///   Otherwise, it is the same as `self.capacity`.
+  /// - Postcondition: if `moveElements` is `true`, `self.count == 0`
+  ///
+  /// - Complexity: O(N).
+  @inline(never)
```
Comment (saeta) [inline Sources/PenguinStructures/ArrayStorage.swift:199]: Could you add a comment explaining why this is `@inline(never)`?

File: Sources/PenguinStructures/ArrayStorage.swift [hunk 3]
```diff
@@ -156,6 +247,21 @@ extension ArrayStorageImplementation {
...
 
+  /// Returns a copy of `self` after appending the instance of the concrete
+  /// element type whose address is `p`, moving elements from the existing
+  /// storage iff `moveElements` is true.
+  ///
+  /// - Postcondition: if `count == capacity` on invocation, the result's
+  ///   `capacity` is `self.capacity` scaled up by a constant factor.
+  ///   Otherwise, it is the same as `self.capacity`.
+  /// - Postcondition: if `moveElements` is `true`, `self.count == 0`
+  ///
```
Comment (saeta) [inline Sources/PenguinStructures/ArrayStorage.swift:258]: Optional:

```suggestion
```

File: Tests/PenguinStructuresTests/ArrayStorageTests.swift [hunk 4]
```diff
@@ -141,7 +236,11 @@ class ArrayStorageTests: XCTestCase {
     ("test_create", test_create),
     ("test_append", test_append),
     ("test_typeErasedAppend", test_typeErasedAppend),
+    ("test_appending", test_appending),
+    ("test_typeErasedAppending", test_typeErasedAppending),
+    ("test_replacementStorage", test_replacementStorage),
```
Comment (saeta) [inline Tests/PenguinStructuresTests/ArrayStorageTests.swift:241]: Nit: it's nice to keep these in the same order as the tests are in the source file.

File: Tests/PenguinStructuresTests/ArrayStorageTests.swift [hunk 5]
```diff
@@ -78,6 +78,80 @@ extension ArrayStorageImplementation where Element: Equatable {
...
+      
+      s1.withUnsafeMutableBufferPointer { b1 in
+        s2.withUnsafeMutableBufferPointer { b2 in
+          XCTAssert(b1.elementsEqual(b2))
+        }
+      }
+    }
+  }
+
+  static func test_replacementStorage<Source: Collection>(source: Source)
```
Comment (saeta) [inline Tests/PenguinStructuresTests/ArrayStorageTests.swift:133]: Nit: starting with `test_` makes it seem to me that this is a test function that happens to be in the wrong place. Please consider some other prefix (e.g. verify).

Comment (dabrahams) [reply Tests/PenguinStructuresTests/ArrayStorageTests.swift:133]: I don't understand the objection.  It *is* a test function; you do see the XCTAssert calls in there.  Don't know what you mean by "the wrong place".  Also, this is consistent with other names in the file.  Are you suggesting that they all be changed?

Comment (dabrahams) [reply Tests/PenguinStructuresTests/ArrayStorageTests.swift:133]: @saeta bump

File: Tests/PenguinStructuresTests/ArrayStorageTests.swift (hunk 4 above)
Comment (dabrahams) [reply Tests/PenguinStructuresTests/ArrayStorageTests.swift:241]: That seems like a bunch of busywork for little benefit, but I will do it if you think it's a good policy.  Should we also alphabetize them? (0.5 😉 )

Comment (dabrahams) [reply Tests/PenguinStructuresTests/ArrayStorageTests.swift]: Done

File: Sources/PenguinStructures/ArrayStorage.swift (hunk 2 above)
Comment (dabrahams) [reply Sources/PenguinStructures/ArrayStorage.swift]: Done

**Pull request merged**